src/components/OfflineWithFeedback.js

```diff
@@ -28,6 +28,12 @@ const propTypes = {
     // eslint-disable-next-line react/forbid-prop-types
     errors: PropTypes.object,
 
+    /** Predetermine whether there are errors, so that we don't have to pass all the errors if we aren't rendering them */
```

Comment:
Ah, is there a problem if we pass the errors and we're not rendering them? Since `hasErrors` is derived from the prop `errors` I think it's fine to pass `errors`. My only concern is that we can introduce a bug if someone uses both props in the wrong way, i.e. `hasErrors={false}` and `errors={objectWithErrors}`

Reply:
> is there a problem if we pass the errors and we're not rendering them?
> i.e. `hasErrors={false}` and `errors={objectWithErrors}`

Thanks for bringing that up, it's interesting. In that case OfflineWithFeedback will internally create `hasErrors` based on `props.errors` and so it will override `props.hasErrors`. 

https://github.com/Expensify/App/blob/170f99d68986e45674eb19504463defc80fd9643/src/components/OfflineWithFeedback.js#L84

To me that makes sense since the intention with the `hasErrors` prop is to be able to determine whether the wrapped component represents something with errors. For example, [here](https://github.com/Expensify/App/blob/170f99d68986e45674eb19504463defc80fd9643/src/components/LHNOptionsList/OptionRowLHN.js#L106) I use `option.brickRoadIndicator` to decide if there are errors in the report. It's determined ahead of time, in an efficient manner, when creating the option [here](https://github.com/Expensify/App/blob/170f99d68986e45674eb19504463defc80fd9643/src/libs/SidebarUtils.js#L217), so it didn't make sense to me to pass all the error messages through.

cc @iwiznia since he created OfflineWithFeedback and @tgolen since he refactored the LHN.

Reply:
> For example, [here](https://github.com/Expensify/App/blob/170f99d68986e45674eb19504463defc80fd9643/src/components/LHNOptionsList/OptionRowLHN.js#L106) I use

This link is broken, can you pass a good one?

Reply:
> doesn't make sense to me to pass all the error messages through

OfflineWithFeedback needs the messages to print them.
Also, the previous code had no hasErrors prop, it was a variable added for easier readability. The fact that you can pass a "wrong" configuration makes this change worse than what it was IMO, but I am not sure the intention of the change... did we find that iterating the errors object passed to OfflineWithFeedback was slow?

Reply:
> This link is broken

Ah shoot. Here's that whole section again with working links.

To me that makes sense since the intention with the `hasErrors` prop is to be able to determine ahead of time whether the wrapped component represents something with errors. For example, [here](https://github.com/Expensify/App/blob/a545d800062b4a2d5c31c003ca0993b3a022102a/src/components/LHNOptionsList/OptionRowLHN.js#L106) I use `option.brickRoadIndicator` to decide if there are errors in the LHN report option row. It's determined ahead of time, in an efficient manner, when creating the option [here](https://github.com/Expensify/App/blob/a545d800062b4a2d5c31c003ca0993b3a022102a/src/libs/SidebarUtils.js#L217), so it doesn't make sense to me to pass all the error messages through.



> OfflineWithFeedback needs the messages to print them.

Option rows in the LHN and search represent reports, and we want them to have 50% opacity when pending creation and when creation has failed. To determine if creation has failed we have to look at any possible error field on the report as well as all of the report actions, so we use the function [getBrickRoadIndicatorStatusForReport](https://github.com/Expensify/App/blob/a545d800062b4a2d5c31c003ca0993b3a022102a/src/libs/OptionsListUtils.js#L219) to return as soon as the first error is found.

We don't want to show error messages on these rows because they are shown in the report instead, so I pass `shouldShowErrorMessages={false}` to prevent them from rendering. I think that's much better than putting a display none on the error row container.

I know there are other ways to do this, but I think the above explanation makes a strong case for this solution. I could have duplicated some of the OfflineWithFeedback logic for the option row, but we also wrapped the ReportScreen header individually in OfflineWithFeedback [here](https://github.com/Expensify/App/blob/a545d800062b4a2d5c31c003ca0993b3a022102a/src/pages/home/ReportScreen.js#L221-L230), and I think we may have more cases that need offline feedback and don't need to show error messages.

Reply:
Ah, I see, so it's for when we show the dot, but not the errors themselves. I can get behind on allowing to pass _either_ hasErrors OR errors (but not both). Can we do that via props validation?

Reply:
Ok that would make sense. I'll see if that's possible.

Reply:
Cool, it looks like you can do this with a [custom prop validator](https://reactjs.org/docs/typechecking-with-proptypes.html#react.proptypes:~:text=//%20You%20can%20also%20specify%20a%20custom%20validator.%20It%20should%20return%20an%20Error%0A%20%20//%20object%20if%20the%20validation%20fails.%20Don%27t%20%60console.warn%60%20or%20throw%2C%20as%20this%0A%20%20//%20won%27t%20work%20inside%20%60oneOfType%60.%0A%20%20customProp%3A%20function(props%2C%20propName%2C%20componentName)%20%7B) function. I'll work on that solution.

Reply:
As I'm working on the custom prop validation, I'm wondering if it's really worth enforcing. It feels like an anti-pattern. I would rather make sure that the component always renders properly, and logically, based on any combination of its props

> I can get behind on allowing to pass either hasErrors OR errors (but not both).

What is wrong with that? If we pass `hasErrors=true` and `errors` is set to an object full of errors, then there aren't any conflicts; the component will render the errors correctly.

If we pass `hasErrors=false` and `errors` is set to an object full of errors, then OfflineWithFeedback will understand that there are in fact errors and it will render them.

If we pass `hasErrors=false` and `erros={}` then we will correctly not render any errors.

If we pass `hasErrors=true` and `errors={}`, then we have a problem because `shouldShowErrorMessages` defaults to `true` and we will render the close button without any error message based on the condition below. I think that might be the problematic case you intended to point out.

https://github.com/Expensify/App/blob/a545d800062b4a2d5c31c003ca0993b3a022102a/src/components/OfflineWithFeedback.js#L104

To remedy that I will add `&& !_.isEmpty(props.errors))`, so if someone tries to indicate that there are errors and doesn't pass the messages, and forgets to set `shouldShowErrorMessages=false`, then the component still renders well.

Reply:
>If we pass hasErrors=false and errors is set to an object full of errors, then OfflineWithFeedback will understand that there are in fact errors and it will render them.

I think this behavior can't be intuitive if this condition happens. What if alternatively use only the prop `shouldShowErrorMessages` to conditionally render the error messages [here](https://github.com/Expensify/App/blob/5e86ae22483bdb060e278bc62152bb4ee6260df9/src/components/OfflineWithFeedback.js#L105)? if `shouldShowErrorMessages=false` and `errors` exist then, we only show the dot and we gracefully set `hasErrors` [here](https://github.com/Expensify/App/blob/5e86ae22483bdb060e278bc62152bb4ee6260df9/src/components/OfflineWithFeedback.js#L84) for the opacity.

Reply:
> I think this behavior isn't intuitive if this condition happens.

Would it be intuitive if I renamed `hasErrors` to `alreadyHasError` or`hasAnError` or `representsAnError`?



> What if alternatively use only the prop `shouldShowErrorMessages` to conditionally render the error messages [here](https://github.com/Expensify/App/blob/5e86ae22483bdb060e278bc62152bb4ee6260df9/src/components/OfflineWithFeedback.js#L105)?

If we took that approach then we would have to update `shouldShowErrorMessages` from the parent component when the errors change, and that would create a lot of duplicate code I think.

Reply:
> wondering if it's really worth enforcing. It feels like an anti-pattern

Why is it not worth enforcing? 
I agree it feels like an anti pattern, but the solution for that is not use the pattern (ie: not have 2 ways of doing the same thing in one component).